### PR TITLE
fleet: implement buckup plugin

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

/kind cleanup
/kind feature

**What this PR does / why we need it**:

support for installing Velero in multi-cluster as a union by applying `Fleet`, which involves setting up the backup plugin.

part of #374 

the api implement of #378 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
support union install velero by apply fleet whoes backup plugin is set.

```

### CODE REVIEW

File: pkg/fleet-manager/fleet_plugin_backup.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	"kurator.dev/kurator/pkg/fleet-manager/plugin"
+	"kurator.dev/kurator/pkg/infra/util"
+)
+
+const (
+	AccessKey = "access-key"
+	SecretKey = "secret-key"
+)
+
+func (f *FleetManager) reconcileVeleroPlugin(ctx context.Context, fleet *fleetv1a1.Fleet, fleetClusters map[ClusterKey]*fleetCluster) (kube.ResourceList, ctrl.Result, error) {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin_backup.go:37]: reconcileBackupPlugin

File: pkg/fleet-manager/plugin/testdata/backup/custom-values.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      repository: velero/velero
+      tag: v1.10.1
+    initContainers:
+    - image: velero/velero-plugin-for-aws:v1.7.1
+      name: velero-plugin-for-aws
+      volumeMounts:
+      - mountPath: /target
+        name: plugins
+    snapshotsEnabled: false
+
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/plugin/testdata/backup/custom-values.yaml:72]: remove

**Review by hzxuzhonghu — COMMENTED** [review]
Wondering reconcile the plugins concurrently, because reconcile them one by one maybe too time costing

**Comment by Xieql** [comment]
> Wondering reconcile the plugins concurrently, because reconcile them one by one maybe too time costing

yes, currentlly, each plugin need wait for velero helm release to be reconciled

File: pkg/fleet-manager/plugin/testdata/backup/custom-values.yaml (hunk 2 above)
Comment (Xieql) [reply pkg/fleet-manager/plugin/testdata/backup/custom-values.yaml:72]: do you means the blank line.
if this blank line(the last line in `values`) is removed, the test will not pass.
see https://github.com/kurator-dev/kurator/blob/772b59656c0deb1e228fc5b8fb0c468908c7da3d/pkg/fleet-manager/plugin/testdata/kyverno/default.yaml#L46
and https://github.com/kurator-dev/kurator/blob/772b59656c0deb1e228fc5b8fb0c468908c7da3d/pkg/fleet-manager/plugin/testdata/prometheus/with-values.yaml#L63

File: pkg/fleet-manager/fleet_plugin_backup.go (hunk 1 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin_backup.go:37]: ok

File: pkg/fleet-manager/plugin/plugin.go [hunk 3]
```diff
@@ -213,6 +216,77 @@ func RendPrometheus(fsys fs.FS, fleetName types.NamespacedName, fleetRef *metav1
...
+					},
+				},
+			},
+		},
+		"credentials": map[string]interface{}{
+			"secretContents": map[string]interface{}{
+				"cloud": fmt.Sprintf(
+					"[default]\naws_access_key_id=%s\naws_secret_access_key=%s",
+					accessKey,
+					secretKey,
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/plugin/plugin.go:264]: We should not allow ak sk used like this, they should be referenced from a secret

File: pkg/fleet-manager/manifests/plugins/velero.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+targetNamespace: velero
+values:
+  image:
+    repository: velero/velero
+    tag: v1.11.1
+  deployNodeAgent: true
+  defaultVolumesToFsBackup: true
+  snapshotsEnabled: false
+  initContainers:
+    - image: velero/velero-plugin-for-aws:v1.7.1
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/manifests/plugins/velero.yaml:14]: what does this init container do

File: pkg/fleet-manager/plugin/plugin.go (hunk 3 above)
Comment (Xieql) [reply pkg/fleet-manager/plugin/plugin.go:264]: ok，we can just pass the secretName

File: pkg/fleet-manager/manifests/plugins/velero.yaml (hunk 4 above)
Comment (Xieql) [reply pkg/fleet-manager/manifests/plugins/velero.yaml:14]: The initContainers configuration is the tool for communicating with AWS S3 or similar storage solutions, like the Minio setup we are using. 
see https://github.com/vmware-tanzu/helm-charts/blob/7564a60a891199aff32dc4f34ca9f1e882e18e59/charts/velero/values.yaml#L58-L60

File: pkg/fleet-manager/plugin/testdata/backup/custom-values.yaml (hunk 2 above)
Comment (Xieql) [reply pkg/fleet-manager/plugin/testdata/backup/custom-values.yaml:72]: we can try to remove it

File: pkg/apis/fleet/v1alpha1/types.go [hunk 5]
```diff
@@ -283,26 +283,49 @@ type BackupStorage struct {
...
+	// Provider specifies the storage provider type (e.g., aws, huaweicloud, gcp, azure).
 	Provider string `json:"provider"`
 	// Endpoint provides the endpoint URL for the storage.
 	Endpoint string `json:"endpoint"`
 	// Region specifies the region of the storage.
-	Region string `json:"region"`
+	// +optional
+	Region string `json:"region,omitempty"`
 	// Config is a map for additional provider-specific configurations.
+	// TODO: enable Config, here is what user can set:
```
Comment (hzxuzhonghu) [inline pkg/apis/fleet/v1alpha1/types.go:313]: TODO?

File: pkg/fleet-manager/fleet_plugin.go [hunk 6]
```diff
@@ -47,30 +48,33 @@ func (f *FleetManager) reconcilePlugins(ctx context.Context, fleet *fleetapi.Fle
...
-			result, ctrlResult, err := f.reconcileBackupPlugin(ctx, fleet, fleetClusters)
-			resultsChannel <- reconcileResult{result, ctrlResult, err}
+			wg.Wait()
+			close(resultsChannel)
 		}()
 
-		for i := 0; i < 4; i++ {
-			res := <-resultsChannel
+		for res := range resultsChannel {
 			if res.err != nil || res.ctrlResult.RequeueAfter > 0 {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin.go:78]: We should aggregate the error result here

**Review by hzxuzhonghu — COMMENTED** [review]
Could you add more test cases

File: pkg/fleet-manager/fleet_plugin.go (hunk 6 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin.go:78]: ok

**Comment by Xieql** [comment]
> Could you add more test cases

ok

**Comment by Xieql** [comment]
still in draft.
the draft label will be removed when this pr is ready

**Comment by Xieql** [comment]
PTAL @hzxuzhonghu

File: pkg/fleet-manager/fleet_plugin.go [hunk 7]
```diff
@@ -35,29 +37,68 @@ const (
 	PrometheusThanosServiceName = "prometheus-prometheus-thanos"
 
 	NoneClusterIP = "None"
+
+	FleetPluinLabel = "fleet.kurator.dev/fleet-plugin"
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin.go:41]: plugin typo

File: pkg/fleet-manager/fleet_plugin_backup.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+// It generates and applies Velero Helm configurations based on the specified backup plugin settings in the fleet specification.
+func (f *FleetManager) reconcileBackupPlugin(ctx context.Context, fleet *v1alpha1.Fleet, fleetClusters map[ClusterKey]*fleetCluster) (kube.ResourceList, ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if fleet.Spec.Plugin.Backup == nil {
+		// reconcilePluginResources will delete all resources if plugin is nil
+		return nil, ctrl.Result{}, nil
+	}
+
+	veleroCfg := fleet.Spec.Plugin.Backup
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin_backup.go:62]: niot: move before L57

File: pkg/fleet-manager/fleet_plugin_backup.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			"cloud": []byte(fmt.Sprintf("[default]\naws_access_key_id=%s\naws_secret_access_key=%s", accessKey, secretKey)),
+		},
+	}
+	return newSecret, nil
+}
+
+func (f *FleetManager) buildHuaWeiCloudSecret(ctx context.Context, secretName string, fleetNN types.NamespacedName) (*corev1.Secret, error) {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin_backup.go:174]: IIRC, this is duplicate with AWS secret

File: pkg/fleet-manager/fleet_plugin_backup.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	AccessKey = "access-key"
+	SecretKey = "secret-key"
+
+	AWS         = "aws"
+	HuaWeiCloud = "huaweicloud"
+	GCP         = "gcp"
+	Azure       = "azure"
+
+	AWSObjStoreSecretName         = "kurator-velero-s3"
+	HuaWeiCloudObjStoreSecretName = "kurator-velero-obs"
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin_backup.go:45]: do we need fixed name? Not quite sure we used fleet name + plugin name before

File: pkg/fleet-manager/fleet_plugin.go [hunk 11]
```diff
@@ -35,29 +37,68 @@ const (
...
+			return ctrl.Result{}, fmt.Errorf("encountered multiple errors: %v", errs)
 		}
-		resources = append(resources, result...)
 
-		result, ctrlResult, err = f.reconcileKyvernoPlugin(ctx, fleet, fleetClusters)
-		if err != nil || ctrlResult.RequeueAfter > 0 {
-			return ctrlResult, err
+		if len(ctrlResults) > 0 {
+			// Handle multiple ctrlResults
+			return ctrlResults[0], nil // TODO: if we need strategy about RequeueAfter
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin.go:100]: Yes, i think this is not right.

If there are > 1 plugins, we only check the first one. If the 1st installed successfully, but the 2nd is being installed, we may wrongly treat all plugins ready,

File: pkg/fleet-manager/fleet_plugin_backup.go (hunk 10 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin_backup.go:45]: it seems that we never create a secret in fleet cluster, while we can add  fleet name + plugin name infomation.

File: pkg/fleet-manager/fleet_plugin_backup.go (hunk 9 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin_backup.go:174]: yes, it can be simplified.

File: pkg/fleet-manager/fleet_plugin_backup.go (hunk 8 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin_backup.go:62]: ok

File: pkg/fleet-manager/fleet_plugin.go (hunk 7 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin.go:41]: ok

File: pkg/fleet-manager/fleet_plugin.go [hunk 12]
```diff
@@ -38,7 +38,7 @@ const (
 
 	NoneClusterIP = "None"
 
-	FleetPluinLabel = "fleet.kurator.dev/fleet-plugin"
+	FleetPluinLabel = "fleet.kurator.dev/plugin-typo"
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/fleet_plugin.go:41]: `typo`

**Review by hzxuzhonghu — APPROVED** [review]

File: pkg/fleet-manager/fleet_plugin.go (hunk 12 above)
Comment (Xieql) [reply pkg/fleet-manager/fleet_plugin.go:41]: I found this name already been defined
see https://github.com/kurator-dev/kurator/blob/5734bb72d2d0a2b8fc00d974e41da5f17560240a/pkg/fleet-manager/util.go#L31C1-L31C1
so I desire to remove it

**Review by hzxuzhonghu — COMMENTED** [review]
/lgtm

**Pull request merged**